### PR TITLE
GC: Implement API

### LINKED ISSUE #377: GC: Implement endpoint to access GC outside of storage engine
Remove data by external invocations via endpoint -- is the simplest case for storage engine garbage collector. Endpoint should provide public method from storage engine to physically remove any existing object. Later this endpoint will be used by `Policer` component and by external monitors via `Control API`.

On delete invocation, endpoint should:
- find shard that stores the object,
- make a GCMARK in graveyard bucket.

To implement this we need:
- [ ] modify `metabase.Ihume()` to support `GCMARK` records

### LINKED ISSUE #378: GC: Run object remover job per shard
Every shard in storage engine should have several monitoring / control routines. One of these jobs: object remover.

Object remover is simple. It iterates over `Graveyard` bucket in shard's metabase and looks for records with `GCMARK` in the value. When worker sees such record, it accumulates key of this record in buffer. When the buffer is full, routine invokes `shard.Delete()` method on these objects.

To implement this we need:
- [ ] provide graveyard iterator in metabase and shard interface,
- [ ] create environment for per-shard routines in storage engine.

### PR DESCRIPTION
**Pull request by cthulhu-rider**
Closes #377.
Closes #378.

### CODE REVIEW

File: pkg/local_object_storage/shard/gc.go [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+		v.prevGroup.Wait()
+
+		var ctx context.Context
+		ctx, v.cancelFunc = context.WithCancel(context.Background())
+
+		v.prevGroup.Add(len(v.handlers))
+
+		for _, h := range v.handlers {
+			err := gc.workerPool.Submit(func() {
+				h(ctx, event)
```
Comment (alexvanin) [inline pkg/local_object_storage/shard/gc.go:128]: `Submit` function eventually creates new go routine. Therefore it is not so good to reuse `h` literal inside. For example this code would fire go vet error.
```go
		for _, h := range v.handlers {
			go func() {
				h(ctx, event)
				v.prevGroup.Done()
			}()
		}
```

I propose to define separate variable for handler pointer:
```go
		for i := range v.handlers {
			h := v.handlers[i]
			err := gc.workerPool.Submit(func() {
				h(ctx, event)
				v.prevGroup.Done()
			})
		}
```

File: pkg/local_object_storage/metabase/util.go [hunk 2]
```diff
@@ -97,6 +108,12 @@ func addressKey(addr *object.Address) []byte {
 	return []byte(addr.String())
 }
 
+// parses object address formed by addressKey.
+func addressFromKey(k []byte) (*object.Address, error) {
```
Comment (alexvanin) [inline pkg/local_object_storage/metabase/util.go:112]: Would be nice to reuse this function in:
- pkg/local_object_storage/metabase/select.go:131
- pkg/local_object_storage/metabase/select.go:439
- pkg/local_object_storage/metabase/iterators.go:134
- pkg/local_object_storage/metabase/movable.go:134

File: pkg/local_object_storage/shard/gc.go [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+func (s *Shard) removeGarbage() {
+	buf := make([]*object.Address, 0, s.rmBatchSize)
+
+	// iterate over metabase graveyard and accumulate
+	// objects with GC mark
+	err := s.metaBase.IterateOverGraveyard(func(g *meta.Grave) error {
+		if g.WithGCMark() {
+			buf = append(buf, g.Address())
+		}
+
```
Comment (alexvanin) [inline pkg/local_object_storage/shard/gc.go:164]: Don't we need to check the length of `buf` there? If `buf` has `s.rmBarchSize` elements, then stop iterating and process buffer.

File: pkg/local_object_storage/shard/gc.go (hunk 1 above)
Comment (cthulhu-rider) [reply pkg/local_object_storage/shard/gc.go:130]: Fixed

File: pkg/local_object_storage/shard/gc.go (hunk 3 above)
Comment (cthulhu-rider) [reply pkg/local_object_storage/shard/gc.go:166]: Definitely yes, fixed

File: pkg/local_object_storage/metabase/iterators.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+		}
+
+		return b.ForEach(func(expKey, _ []byte) error {
+			bktExpired := b.Bucket(expKey)
+			if bktExpired == nil {
+				return nil
+			}
+
+			return bktExpired.ForEach(func(idKey, _ []byte) error {
+				expiresAt, err := strconv.ParseUint(string(expKey), 10, 64)
```
Comment (alexvanin) [inline pkg/local_object_storage/metabase/iterators.go:62]: We don't need to calculate `expiresAt` and make an epoch check on each record in `bktEpxired` bucket. We can move it higher:

```go
			bktExpired := b.Bucket(expKey)
			if bktExpired == nil {
				return nil
			}
			expiresAt, err := strconv.ParseUint(string(expKey), 10, 64)
			...
```

File: pkg/local_object_storage/metabase/util.go (hunk 2 above)
Comment (cthulhu-rider) [reply pkg/local_object_storage/metabase/util.go:112]: Done

File: pkg/local_object_storage/metabase/iterators.go (hunk 4 above)
Comment (cthulhu-rider) [reply pkg/local_object_storage/metabase/iterators.go]: Done

**Review by alexvanin — APPROVED** [review]

**Pull request merged**